rust/lance/src/dataset/scanner.rs

```diff
@@ -2708,14 +2717,44 @@ impl Scanner {
 
         let (match_plan, flat_match_plan) = match &index {
             Some(index) => {
+                // Check if any of the scanner's fragments are covered by the FTS index
+                let indexed_frags_bitmap = index
+                    .fragment_bitmap
+                    .as_ref()
+                    .map(|bitmap| bitmap & &self.get_fragments_as_bitmap())
+                    .unwrap_or_else(|| self.get_fragments_as_bitmap());
+
+                let has_indexed_fragments = !indexed_frags_bitmap.is_empty();
+
+                if !has_indexed_fragments {
+                    // No indexed fragments in the scan - do flat search only on scanner's fragments
+                    let fragments_to_search = if let Some(frags) = &self.fragments {
+                        frags.clone()
+                    } else {
+                        self.dataset.fragments().iter().cloned().collect()
+                    };
+                    let flat_match_plan = self
+                        .plan_flat_match_query(fragments_to_search, query, params, filter_plan)
+                        .await?;
+                    return Ok(flat_match_plan);
+                }
+
                 let match_plan: Arc<dyn ExecutionPlan> = Arc::new(MatchQueryExec::new(
                     self.dataset.clone(),
                     query.clone(),
                     params.clone(),
                     prefilter_source.clone(),
                 ));
 
-                let unindexed_fragments = self.dataset.unindexed_fragments(&index.name).await?;
+                let mut unindexed_fragments = self.dataset.unindexed_fragments(&index.name).await?;
+
+                // Filter by scanner's fragment list to respect fragment restrictions
+                if let Some(scanner_frags) = &self.fragments {
+                    let scanner_ids: std::collections::HashSet<_> =
```

Comment:
we can use `self.get_fragments_as_bitmap();` in order to perform the same operation instead of using a hashset.